proper sharing in ShareExternalPointer

jerryzh168 opened the pull request:
Summary:
Make ShareData and ShareExternalPointer to create new storage when the old one is used by multiple tensors.
When we need to modify the field of storage, we'll create a new storage instead.

Differential Revision: D9350686

ezyang commented on caffe2/core/storage.h:
```diff
@@ -32,6 +35,35 @@ class StorageImpl {
+    CAFFE_ENFORCE_WITH_CALLER(
+        data_type_.id() != TypeIdentifier::uninitialized(),
+        "To create storage with a raw external pointer you need to have meta "
+        "already initialized.");
```
I mentioned this previously, but "meta already initialized" is misleading for a constructor; you want to refer to the meta that is passed in.

ezyang commented on caffe2/core/storage.h:
```diff
@@ -32,6 +35,35 @@ class StorageImpl {
+        "already initialized.");
+    // Check if the deleter is a MemoryDeleter and is a simple nullptr.
+    if (std::is_same<MemoryDeleter, Deleter>::value &&
+        reinterpret_cast<MemoryDeleter*>(&d)[0] == nullptr) {
```
I think this code will trip some strict aliasing Werror builds internally? (May be safer to cast to `void*` first, and then to `MemoryDeleter*`)

ezyang commented on caffe2/core/storage.h:
```diff
@@ -43,14 +75,22 @@ class StorageImpl {
+    return data_ptr_;
+  }
+
+  const DataPtr& data_ptr() const {
```
Thanks, convergences like this are much appreciated. And please feel free to push back if you think the API should look a little different; we were a lot less strict about API of Storage C++ side.

ezyang commented on caffe2/core/storage.h:
```diff
@@ -63,8 +103,8 @@ class StorageImpl {
-  inline void set_device_type(DeviceType device_type) {
-    device_type_ = device_type;
+  void set_numel(int64_t numel) {
+    capacity_ = numel * itemsize();
```
Is this necessary?

jerryzh168 replied:
This is added temporarily to remove friending Tensor in Storage. We'll probably move some of the implementations into Storage in the future so we don't need to set capacity_ outside of Storage.

ezyang replied:
OK. Add it as a work item in "nice to do later" and we can keep it.

ezyang commented on caffe2/core/storage.h:
```diff
@@ -95,26 +141,43 @@ class StorageImpl {
-  // is ready
-  using DataPtr = std::shared_ptr<void>;
+  void ShareData(const Storage& src) {
+    data_ptr_ = src->data_ptr_;
```
This isn't going to work when `DataPtr` becomes a unique pointer like thing. Is this function really the one you wanted?

jerryzh168 replied:
umm.. we can't do `storage_ = src.storage();` since that doesn't keep the old semantics. can DataPtr be shared_ptr?

ezyang replied:
In offline conversation, we concluded that this would be OK if (1) Resize() detaches the storage when it reallocates (which it does) and (2) Storages don't change data_type inplace (which they don't)

ezyang commented on caffe2/core/tensor.h:
```diff
@@ -470,10 +472,13 @@ class CAFFE2_API Tensor {
-        "Can't share data if underlying storage used by more than one tensor");
-    storage_->data_type_ = src.storage()->data_type_;
+    // Right now, we are assuming the device_type are the same, since it is
+    // inheriently the same in the non-templatized code We should probably add
```
nit: inherently, and also missing period

ezyang commented on caffe2/core/storage.h:
```diff
@@ -32,6 +35,35 @@ class StorageImpl {
+      TypeMeta data_type,
+      void* src,
+      TIndex size,
+      size_t capacity,
```
Should document the difference between size and capacity prominently.

jerryzh168 replied:
updated the diff with size removed.

jerryzh168 committed:
proper sharing in ShareExternalPointer (#10804)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10804

Make ShareData and ShareExternalPointer to create new storage when the old one is used by multiple tensors.
When we need to modify the field of storage, we'll create a new storage instead.

Differential Revision: D9350686

fbshipit-source-id: 86445136a22cadedee5e3ebe56073e7d3177dc66
